rust/ip-packet/src/lib.rs

```diff
@@ -885,7 +885,46 @@ impl IpPacket {
         }
     }
 
-    pub fn with_ecn(mut self, ecn: Ecn) -> Self {
+    /// Updates the ECN flags of this packet with the ECN value from the transport layer.
+    ///
+    /// After tunneling an IP packet, we need to merge the ECN flags from the transport layer with the ones already set on the IP packet.
+    /// Essentially, the only time we need to update the ECN flags is:
+    /// - if the IP packet signalled an ECN-capable transport (i.e. the originating network stack is ECN-capable)
+    /// - and we have experienced congestion along the way (i.e. the provided ECN value is [`Ecn::Ce`]).
+    pub fn with_ecn_from_transport(self, ecn: Ecn) -> Self {
+        let ecn = match (self.ecn(), ecn) {
+            (Ecn::NonEct, Ecn::NonEct)
+            | (Ecn::Ect1, Ecn::Ect1)
+            | (Ecn::Ect0, Ecn::Ect0)
+            | (Ecn::Ce, Ecn::Ce) => {
+                // No change needed
+                return self;
+            }
+            (Ecn::NonEct, Ecn::Ect0 | Ecn::Ect1 | Ecn::Ce) => {
+                // Packet sender is not ECN-capable, ignore any update.
+                return self;
+            }
+            (Ecn::Ect1 | Ecn::Ect0 | Ecn::Ce, Ecn::NonEct) | (Ecn::Ce, Ecn::Ect0 | Ecn::Ect1) => {
+                // We already have ECN information, refuse to clear it.
+                return self;
+            }
+            (Ecn::Ect1, Ecn::Ect0) | (Ecn::Ect0, Ecn::Ect1) => {
+                // Don't switch between ECT0 and ECT1, they are equivalent.
+                return self;
+            }
+            (Ecn::Ect1, Ecn::Ce) | (Ecn::Ect0, Ecn::Ce) => {
+                // ECN-capable transport has been signalled and our update is CE, update it!
```

Comment:
Minor suggestion, feel free to ignore, but would it be slightly easier to follow this logic if these flags were grouped as booleans named after what they mean?

Reply:
I am not sure what you mean? `Ecn` is an enum so it has 4 values, I can't express that as a boolean.

Reply:
I agree that this isn't easy to follow if your eyes are not trained to read to Rust. It does have a big advantage though: The compiler checks for us that we are handling all combinations exhaustively.

The way to read this is to look for tuples and consider the left hand as the ECN bit we have and the right hand the ECN bit we got. So for example:

```
(Ecn::NonEct, Ecn::Ect0 | Ecn::Ect1 | Ecn::Ce) => {
    // Packet sender is not ECN-capable, ignore any update.
    return self;
}
```

This means, if we have `Ecn::NonEct` and we get either `Ect0`, `Ect1` or `Ce`, then ignore the update. `NonEct` means non-ECN compatible transport so there is no point to "ugprading" it by flipping any of the bits.

Reply:
I've added a glob import for the enum values, that should reduce the noise and make this easier to read.